File: urc_scripts/urc_scripts/joy_drive.py
import rclpy
from rclpy.publisher import Publisher
from rclpy.subscription import Subscription
from rclpy.node import Node
from rclpy.parameter import Parameter
from rcl_interfaces.msg import SetParametersResult
from geometry_msgs.msg import TwistStamped
from sensor_msgs.msg import Joy


class JoyDrive(Node):
    """
    Use joystick to command the drivetrain.

    Args:
        Node (_type_): node from ros2.
    """

    def __init__(self):
        super().__init__("joy_drive")

        # declare parameters
        self.declare_parameter("max_linear_velocity_ms", 0.0)
        self.declare_parameter("max_angular_velocity_radians", 0.0)
        self.declare_parameter("joy_command_topic", "/joy")
        self.declare_parameter("cmd_vel_topic", "/cmd_vel")
        self.declare_parameter("target_axes", [1, 3])
        self.declare_parameter("invert_linear_velocity", False)
        self.declare_parameter("invert_angular_velocity", False)

        # get parameters
        self.max_linear = self.get_parameter(
            "max_linear_velocity_ms").get_parameter_value().double_value
        self.max_angular = self.get_parameter(
            "max_angular_velocity_radians").get_parameter_value().double_value
        self.joy_command_topic = self.get_parameter(
            "joy_command_topic").get_parameter_value().string_value
        self.cmd_vel_topic = self.get_parameter(
            "cmd_vel_topic").get_parameter_value().string_value
        self.op_axes = self.get_parameter(
            "target_axes").get_parameter_value().integer_array_value
        self.inv_linvel = self.get_parameter(
            "invert_linear_velocity").get_parameter_value().double_value
        self.inv_angvel = self.get_parameter(
            "invert_angular_velocity").get_parameter_value().double_value

        # register dynamic parameter callbacks
        self.add_on_set_parameters_callback(self.on_param_update)

        self.curr_twist = TwistStamped()
        self.joy_msg_subscriber: Subscription = self.create_subscription(
            Joy, self.joy_command_topic, self.update_vel_target, 10
        )
        self.cmd_vel_publisher: Publisher = self.create_publisher(
            TwistStamped, self.cmd_vel_topic, 10
        )

        self.get_logger().info("Joy Drive Started.")
        self.get_logger().info(
            f"Default Settings: [linear velocity {self.max_linear} m/s],"
            + f"[angular velocity {self.max_angular} rad/s]"
        )

    def update_vel_target(self, msg: Joy):
        """Update velocity target for the drivetrain.

        Args:
            msg (Joy): joy message from the controller.
        """
        self.curr_twist.twist.linear.x = msg.axes[
            self.op_axes[0]
        ] * self.max_linear * (-1 if self.inv_linvel else 1)
        self.curr_twist.twist.angular.z = msg.axes[
            self.op_axes[1]
        ] * self.max_angular * (-1 if self.inv_angvel else 1)
        self.curr_twist.header.stamp = self.get_clock().now().to_msg()

        self.cmd_vel_publisher.publish(self.curr_twist)

    def on_param_update(self, params: list[Parameter]) -> SetParametersResult:
        """When parameter is updated.

        Args:
            params (list[Parameter]): list of updated parameters

        Returns:
            SetParametersResult: result for setting the parameters.
        """
        result = SetParametersResult()

        for param in params:
            if param.name == "max_linear_velocity_ms":
                self.max_linear = param.get_parameter_value().double_value
                self.get_logger().info(
                    f"Max linvel updated to {self.max_linear} m/s."
                )
            if param.name == "max_angular_velocity_radians":
                self.max_angular = param.get_parameter_value().double_value
                self.get_logger().info(
                    f"Max angvel updated to {self.max_angular} rad/s."
                )
            if param.name == "target_axes":
                self.op_axes = param.get_parameter_value().integer_array_value
                self.get_logger().info(
                    f"Axies has be configured to {self.op_axes}."
                )
            if param.name == "invert_linear_vel":
                self.inv_linvel = param.get_parameter_value().bool_value
                self.get_logger().info(
                    f"Linear velocity inversion: {self.inv_linvel}"
                )
            if param.name == "invert_angular_vel":
                self.inv_angvel = param.get_parameter_value().bool_value
                self.get_logger().info(
                    f"Angular velocity inversion: {self.inv_angvel}"
                )
        result.successful = True
        return result


def main(args=None):
    """Entry function

    Args:
        args (_type_, optional): Arguments. Defaults to None.
    """
    rclpy.init(args=args)
    node = JoyDrive()
    while rclpy.ok():
        rclpy.spin_once(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: urc_platform/src/joy_drive.cpp
#include "joy_drive.hpp"
#include <rclcpp/qos.hpp>
#include <string>

namespace joy_drive
{

JoyDrive::JoyDrive(const rclcpp::NodeOptions & options)
: rclcpp::Node("joy_drive", options)
{
  // Declare parameters
  declare_parameter("max_linear_velocity_ms", 0.0);
  declare_parameter("max_angular_velocity_radians", 0.0);
  declare_parameter("joy_command_topic", "/joy");
  declare_parameter("cmd_vel_topic", "/cmd_vel");
  declare_parameter("target_axes", std::vector<int64_t>{1, 3});
  declare_parameter("invert_linear_velocity", false);
  declare_parameter("invert_angular_velocity", false);

  // Get parameters
  max_linear_ = get_parameter("max_linear_velocity_ms").as_double();
  max_angular_ = get_parameter("max_angular_velocity_radians").as_double();
  joy_command_topic_ = get_parameter("joy_command_topic").as_string();
  cmd_vel_topic_ = get_parameter("cmd_vel_topic").as_string();
  auto axes = get_parameter("target_axes").as_integer_array();
  target_axes_ = std::make_pair(axes[0], axes[1]);
  invert_axes_ = std::make_pair(
    get_parameter("invert_linear_velocity").as_bool(),
    get_parameter("invert_angular_velocity").as_bool());

  // Register parameter callback
  param_callback_handle_ = add_on_set_parameters_callback(
    [this](const std::vector<rclcpp::Parameter> & params) {
      return on_param_update(params);
    });

  // Create subscribers and publishers
  joy_subscriber_ = create_subscription<sensor_msgs::msg::Joy>(
    joy_command_topic_, rclcpp::SystemDefaultsQoS(),
    [this](const sensor_msgs::msg::Joy & msg) {joy_callback(msg);});

  cmd_vel_publisher_ = create_publisher<geometry_msgs::msg::TwistStamped>(
    cmd_vel_topic_, rclcpp::SystemDefaultsQoS());

  // Log startup message
  RCLCPP_INFO(get_logger(), "Joy Drive Started.");
  RCLCPP_INFO(
    get_logger(), "Default Settings: [linear velocity %f m/s], [angular velocity %f rad/s]",
    max_linear_, max_angular_);
}

void JoyDrive::joy_callback(const sensor_msgs::msg::Joy & msg)
{
  geometry_msgs::msg::TwistStamped twist;
  twist.twist.linear.x = msg.axes[target_axes_.first] * max_linear_ * (invert_axes_.first ? -1 : 1);
  twist.twist.angular.z = msg.axes[target_axes_.second] * max_angular_ * (invert_axes_.second ? -1 : 1);
  twist.header.stamp = now();

  cmd_vel_publisher_->publish(twist);
}

rcl_interfaces::msg::SetParametersResult JoyDrive::on_param_update(const std::vector<rclcpp::Parameter> & params)
{
  rcl_interfaces::msg::SetParametersResult result;
  result.successful = true;

  for (const auto & param : params) {
    if (param.get_name() == "max_linear_velocity_ms") {
      max_linear_ = param.as_double();
      RCLCPP_INFO(get_logger(), "Max linvel updated to %f m/s.", max_linear_);
    } else if (param.get_name() == "max_angular_velocity_radians") {
      max_angular_ = param.as_double();
      RCLCPP_INFO(get_logger(), "Max angvel updated to %f rad/s.", max_angular_);
    } else if (param.get_name() == "target_axes") {
      auto axes = param.as_integer_array();
      target_axes_ = std::make_pair(axes[0], axes[1]);
      RCLCPP_INFO(get_logger(), "Axes updated to [%ld, %ld].", target_axes_.first, target_axes_.second);
    } else if (param.get_name() == "invert_linear_velocity") {
      invert_axes_.first = param.as_bool();
      RCLCPP_INFO(get_logger(), "Linear velocity inversion: %s", invert_axes_.first ? "true" : "false");
    } else if (param.get_name() == "invert_angular_velocity") {
      invert_axes_.second = param.as_bool();
      RCLCPP_INFO(get_logger(), "Angular velocity inversion: %s", invert_axes_.second ? "true" : "false");
    }
  }

  return result;
}

}  // namespace joy_drive

#include <rclcpp_components/register_node_macro.hpp>
RCLCPP_COMPONENTS_REGISTER_NODE(joy_drive::JoyDrive)